ksql-engine/src/main/java/io/confluent/ksql/function/udaf/KudafUndoAggregator.java

```diff
@@ -0,0 +1,58 @@
+/**
+ * Copyright 2018 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+
+package io.confluent.ksql.function.udaf;
+
+import io.confluent.ksql.GenericRow;
+import io.confluent.ksql.function.KsqlUndoableAggregationFunction;
+import org.apache.kafka.streams.kstream.Aggregator;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+public class KudafUndoAggregator implements Aggregator<String, GenericRow, GenericRow> {
+  private Map<Integer, KsqlUndoableAggregationFunction> aggValToAggFunctionMap;
+  private Map<Integer, Integer> aggValToValColumnMap;
+
+  public KudafUndoAggregator(
+      Map<Integer, KsqlUndoableAggregationFunction> aggValToAggFunctionMap,
+      Map<Integer, Integer> aggValToValColumnMap) {
+    Objects.requireNonNull(aggValToAggFunctionMap);
+    Objects.requireNonNull(aggValToValColumnMap);
+    this.aggValToAggFunctionMap = Collections.unmodifiableMap(
+        new HashMap<>(aggValToAggFunctionMap));
+    this.aggValToValColumnMap = Collections.unmodifiableMap(new HashMap<>(aggValToValColumnMap));
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public GenericRow apply(String s, GenericRow rowValue, GenericRow aggRowValue) {
+    aggValToValColumnMap.forEach(
+        (aggRowIndex, rowIndex) ->
+            aggRowValue.getColumns().set(aggRowIndex, rowValue.getColumns().get(rowIndex)));
+
+    aggValToAggFunctionMap.forEach(
+        (aggRowIndex, function) ->
+            aggRowValue.getColumns().set(
+                aggRowIndex,
+                function.undo(
+                    rowValue.getColumns().get(function.getArgIndexInValue()),
+                    aggRowValue.getColumns().get(aggRowIndex))));
```

Comment:
Not sure I follow what's going on here. The `aggRowValue[aggRowIndex]` is set to `rowValue[rowIndex]` on line 47. 

on this line, we are simply performing an aggregate on the value, which makes me wonder what purpose the operation on line 47 is. Doesn't the operation on line 47 clobber the existing aggregation value? It _seems_ to be setting the aggregation value to the incoming value. Why is it correct?

(I am confident it's correct, I am asking to clarify my understanding.)

I know this pattern is copied over from the existing aggregation function, so it may make sense to have a comment in the original pattern to explain what is going on here.

Reply:
The first statement is copying all the grouping expressions and arguments into the row. The second statement does the aggregation and writes that into the row.